proxy/internal/modules/store/mysql.go

```diff
@@ -0,0 +1,496 @@
+	insertModuleSQL
+	selectRegistryIDSQL
+	selectAllRegistryIDsSQL
+	countVersionsSQL
+	selectModuleRevInfoSQL
+	selectGoModContentsSQL
+	selectModuleVersionsSQL
+	updateRegistryIDSQL
+	deleteModuleSQL
+)
+
+type statements map[int]*sql.Stmt
+
+func load(kind string, db *sql.DB) (statements, error) {
+	loaded := make(statements, len(mySQLTexts))
+
+	stmtTexts := mySQLTexts
+	if kind == "postgres" {
+		return loaded, errors.New("postgres is not supported (issue #103)")
+		// stmtTexts = postgreSQLTexts
+	}
+
+	for id, text := range stmtTexts {
+		stmt, err := db.Prepare(text)
+		if err != nil {
+			return nil, errors.Wrapf(err, "bad sql statement: %q", text)
+		}
+		loaded[id] = stmt
+	}
+
+	return loaded, nil
+}
+
+var (
+	mySQLTexts = map[int]string{
+		// modules
+		insertModuleContentsSQL: `insert into proxy_module_contents(path, contents) values (?, ?)`,
+		selectModuleContentsSQL: `select contents from proxy_module_contents where path=?`,
+		moduleContentsExistsSQL: `select count(id) from proxy_module_contents where path=?`,
+		deleteModuleContentsSQL: `delete from proxy_module_contents where path=?`,
+		// index
+		insertModuleSQL:         `insert into proxy_modules(source, version, go_mod_contents, rev_info_contents, registry_mod_id) values (?, ?, ?, ?, ?)`,
+		selectRegistryIDSQL:     `select registry_mod_id from proxy_modules where source=? and version=?`,
+		selectAllRegistryIDsSQL: `select registry_mod_id from proxy_modules`,
+		countVersionsSQL:        `select count(version) from proxy_modules group by source`,
+		selectModuleRevInfoSQL:  `select rev_info_contents from proxy_modules where source=? and version=?`,
+		selectGoModContentsSQL:  `select go_mod_contents from proxy_modules where source=? and version=?`,
+		selectModuleVersionsSQL: `select version from proxy_modules where source=?`,
+		updateRegistryIDSQL:     `update proxy_modules set registry_mod_id=? where source=? and version=?`,
+		deleteModuleSQL:         `delete from proxy_modules where source=? and version=?`,
+	}
+)
+
+func (m *mysqlStore) insertModulesContents(mod coordinates.Module, blob repository.Blob) error {
+	exists, err := m.modelContentsExists(mod)
+	if err != nil {
+		return err
+	}
+
+	path := pathOf(mod)
+	if exists {
+		m.log.Warnf("not saving %s because we already have it @ %s", mod, path)
+		return errors.Errorf("already have a copy of %s", mod)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), dbTimeout)
+	defer cancel()
+	if _, err = m.statements[insertModuleContentsSQL].ExecContext(ctx, path, []byte(blob)); err != nil {
+		m.emitter.Count("db-insertmodule-failure", 1)
+		m.log.Errorf("failed to write zip for %s, %+v", mod, err)
+	}
+	return err
+}
+
+func (m *mysqlStore) modelContentsExists(mod coordinates.Module) (bool, error) {
+	path := pathOf(mod)
+	start := time.Now()
+	ctx, cancel := context.WithTimeout(context.Background(), dbTimeout)
+	defer cancel()
+	rows, err := m.statements[moduleContentsExistsSQL].QueryContext(ctx, path)
```

Comment:
this should only return one row, right? so lets use
https://golang.org/pkg/database/sql/#Conn.QueryRowContext

Reply:
TIL

Reply:
replaced in all queries that expect only one row